pkg/clusteragent/externalmetrics/model/datadogmetricinternal.go

```diff
@@ -205,3 +219,23 @@ func (d *DatadogMetricInternal) newCondition(status bool, updateTime metav1.Time
 
 	return condition
 }
+
+// resolveQuery tries to resolve the query and set the DatadogMetricInternal fields accordingly
+func (d *DatadogMetricInternal) resolveQuery(query string) {
+	resolvedQuery, resolved, err := resolveQuery(query)
+	if err != nil {
+		log.Errorf("Unable to resolve DatadogMetric query %q: %w", d.Query, err)
+		d.Valid = false
+		d.Error = fmt.Errorf("Cannot resolve query: %v", err)
+		d.UpdateTime = time.Now().UTC()
+		d.resolvedQuery = nil
+		return
+	}
+	if resolved {
+		// Update the query in case it's been changed due to template resolution
+		d.Query = resolvedQuery
```

Comment:
If we do change `d.Query`, I think `UpdateFrom` will always trigger.
Also I think the `resolved == true` bool should be equivalent to `err == nil`.